Refactor CMake to be nicer, more modern, easier to understand

vstumpf opened the pull request:
<!-- NOTE: Anything within these brackets will be hidden on the preview of the Pull Request. -->

* **Addressed Issue(s)**: N/A

<!--
Please specify the rAthena [GitHub issue(s)](https://help.github.com/articles/autolinked-references-and-urls/#issues-and-pull-requests) this pull request amends.
If no issue exists yet, please [create one](https://github.com/rathena/rathena/issues/new) first and then link your pull request to the amendment!
-->

* **Server Mode**: Both

<!-- Which mode does this pull request apply to: Pre-Renewal, Renewal, or Both? -->

* **Description of Pull Request**: 
* Use modern CMake
* Drop a lot of the configuration that we don't really use
  * if anyone does use it, we can add it later
* people can install CMake with Visual Studio, so we wont have to manually work with the sln/vcxproj files anymore
* people can install CMake on Linux (available on all major distros), we won't have to use autotools anymore
* eventually, we can remove the existing build systems and just use CMake

<!-- Describe how this pull request will resolve the issue(s) listed above. -->

CairoLee commented on src/map/CMakeLists.txt:
```diff
@@ -1,45 +1,125 @@
+	"battleground.cpp"
+	"buyingstore.cpp"
+	"cashshop.cpp"
+	"channel.cpp"
+	"chat.cpp"
+	"chrif.cpp"
+	"clan.cpp"
+	"clif.cpp"
+	"date.cpp"
+	"duel.cpp"
+	"elemental.cpp"
+	"guild.cpp"
+	"homunculus.cpp"
+	"instance.cpp"
+	"intif.cpp"
+	"itemdb.cpp"
+	"log.cpp"
+	"mail.cpp"
+	"map.cpp"
+	"mapreg.cpp"
+	"mercenary.cpp"
+	"mob.cpp"
+	"navi.cpp"
+	"npc.cpp"
+	"npc_chat.cpp"
+	"party.cpp"
+	"path.cpp"
+	"pc.cpp"
+	"pc_groups.cpp"
+	"pet.cpp"
+	"quest.cpp"
+	"script.cpp"
+	"searchstore.cpp"
+	"skill.cpp"
+	"status.cpp"
+	"storage.cpp"
+	"trade.cpp"
+	"unit.cpp"
+	"vending.cpp"
+)
```
Consider using the following to automatically introduce all files with cpp suffixes into the MAP_SOURCES variable, so that any future addition of cpp files does not require changes to the CMakeLists.txt file.

`file(GLOB MAP_SOURCES ${MAP_SOURCE_DIR}/*.cpp)`

vstumpf replied:
I specifically do not want to do this, when we add a new file, CMake won't know to regenerate the build system.
See the [official documentation](https://cmake.org/cmake/help/latest/command/file.html?highlight=CONFIGURE_DEPENDS#filesystem)

CairoLee commented on src/map/CMakeLists.txt:
```diff
@@ -1,45 +1,125 @@
+	"chrif.hpp"
+	"clan.hpp"
+	"clif.hpp"
+	"clif_obfuscation.hpp"
+	"clif_packetdb.hpp"
+	"clif_shuffle.hpp"
+	"date.hpp"
+	"duel.hpp"
+	"elemental.hpp"
+	"guild.hpp"
+	"homunculus.hpp"
+	"instance.hpp"
+	"intif.hpp"
+	"itemdb.hpp"
+	"log.hpp"
+	"mail.hpp"
+	"map.hpp"
+	"mapreg.hpp"
+	"mercenary.hpp"
+	"mob.hpp"
+	"navi.hpp"
+	"npc.hpp"
+	"packets.hpp"
+	"packets_struct.hpp"
+	"party.hpp"
+	"path.hpp"
+	"pc_groups.hpp"
+	"pc.hpp"
+	"pet.hpp"
+	"quest.hpp"
+	"script_constants.hpp"
+	"script.hpp"
+	"searchstore.hpp"
+	"skill.hpp"
+	"status.hpp"
+	"storage.hpp"
+	"trade.hpp"
+	"unit.hpp"
+	"vending.hpp"
+)
```
Consider using the following to automatically introduce all files with hpp suffixes into the MAP_HEADERS variable, so that any future addition of hpp files does not require changes to the CMakeLists.txt file.

`file(GLOB MAP_HEADERS ${MAP_SOURCE_DIR}/*.hpp)`

vstumpf commented on src/web/CMakeLists.txt:
```diff
@@ -1,41 +1,42 @@
-if ( NOT ENABLE_WEB_SERVER )
-	return()
-endif ( NOT ENABLE_WEB_SERVER )
+target_sources(web-server PRIVATE
```
As I responded in the above review, this isn't great.
> I specifically do not want to do this. When we add a new file, CMake won't know to regenerate the build system.
See the [official documentation](https://cmake.org/cmake/help/latest/command/file.html?highlight=CONFIGURE_DEPENDS#filesystem)

CairoLee replied:
I still agree with using `file(GLOB...) ` to get the list of files, although he does have the problem @vstumpf mentioned.

But I think if the problem did arise, I would personally use: `rm -rf * && cmake -G "Unix Makefiles" .. && make` in the `cbuild` directory, which is actually very low cost for me.

For a novice developer, `file(GLOB...)' is a lot less thought-provoking than having to find the corresponding CMakeLists.txt and add hpp and cpp respectively when adding files. Is much more thoughtless than having to find the corresponding CMakeLists.txt and find the corresponding location for hpp and cpp respectively.

vstumpf replied:
The chances of someone adding a file are pretty low, it's usually reserved for developers who (should) know what they're doing.
I'm imagining a user who updates rathena every couple months or so; if we change CMakeLists when a new file appears, IDEs (Visual Studio, vscode, etc) will automatically regenerate the build system. Otherwise, they will have to go in and clear the cache to get cmake to rebuild.
Ideally, finding the correct CMakeLists.txt won't be too hard, since it'll be in the same directory as the source file.

vstumpf commented on src/web/CMakeLists.txt:
```diff
@@ -1,41 +1,42 @@
+target_link_libraries(web-server PUBLIC
+	common
+	httplib
+	Threads::Threads
```
This comes from [FindThreads](https://cmake.org/cmake/help/latest/module/FindThreads.html)

vstumpf commented on db/CMakeLists.txt:
```diff
@@ -0,0 +1,72 @@
+
+# Why do we list all the files instead of using glob?
+# This way, if we add a new file, the build system will know to regenerate
+# and the new file will be copied over.
```
No
Honestly, everyone is scared of adding new files anyways, I'll probably be the only one who creates new files.

Lemongrass3110 commented on .github/workflows/build_servers_cmake.yml:
```diff
@@ -31,23 +31,19 @@ jobs:
           # The ubuntu-latest label currently points to ubuntu-22.04.
           # Available: ubuntu-22.04, ubuntu-20.04
-          os: [ubuntu-latest]
+          os: [ubuntu-latest, windows-latest]
```
If cmake (ubuntu) is already used for the other builds, it might be wise to only run windows here.

vstumpf replied:
Renamed this to show it was just windows too

Lemongrass3110 commented on .github/workflows/analysis_codeql.yml:
```diff
@@ -31,8 +31,6 @@ jobs:
           # Available: ubuntu-22.04, ubuntu-20.04
           os: [ubuntu-latest]
-          # Older versions of GCC are not available via unaltered aptitude repo lists.
-          gcc: ['10']
```
By removing this it will use the default compiler, which is most likely newer.
I intentionally set the compiler to a lower version, because they were tons faster during our tests back then.

vstumpf replied:
Checking the actions for this PR, the difference is <10% https://github.com/rathena/rathena/actions/runs/5252590012

Lemongrass3110 commented on tools/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+    )
+
+    copy_to_source_dir("${LINUX_FILES_TO_COPY}")
+endif()
```
Missing empty newline

Lemongrass3110 commented on tools/athena-start:
```diff
@@ -0,0 +1,227 @@
+#!/bin/sh
```
Can we rename it to .sh while we are at it?
And shouldnt the scripts you moved into tools be removed from the root then?

vstumpf replied:
I don't want to break the current workflow for now, but I put a log in the old `athena-start` script telling people to use cmake and `./athena-start.sh` instead

Lemongrass3110 commented on .gitignore:
```diff
@@ -139,6 +139,12 @@ Thumbs.db
 /navigenerator.bat
 
+# Linux script tools
+# /athena-start
```
Should be uncommented I guess

vstumpf commented:
Here's what the "deprecation notice" looks like in rathena.sln
![image](https://github.com/rathena/rathena/assets/6844975/3a35b863-66ef-499c-8918-3b8fa9ab7688)

vstumpf commented:
🙏

Lemongrass3110 commented:
Since you requested a review from me:
Please address the pending comments (mostly CI related).
And I did not read something that we all agreed on wanting to drop MSVS support in the near future?
If the CMake support is as broken as the Clang Format support in MSVS I would rather oppose that.

vstumpf commented:
Visual Studio started supporting CMake in 2017, and it works pretty well: [https://devblogs.microsoft.com/cppblog/cmake-support-in-visual-studio/](Microsoft dev blog from 2016)

This has been brought up in developing features like the web-server and map-server-generator, and discord conversations. The issue is having to update **three** different systems to add a new file/project/library. Because CMake can be used to generate the Visual Studio solution files (or just point VS to the CMakeLists.txt), Unix Makefiles, or even Ninja files, we would just have to make changes to the CMake build system. Write it once, it works everywhere.

vstumpf commented:
> Will it still provide the same benefits for debugging as the current solution does?

A user can open the rathena directory and use the builtin cmake support (which does have the same features) or can download cmake and generate the solution file themselves; then they can use msvs to open the solution file

aleos89 commented:
Sorry, one last question since I don't use CMake much, but does this mean the use of MSVS would go away with this PR? It's my main bread and butter with testing rAthena.

vstumpf commented:
Nope! You can definitely still use msvs. You'll just open the project in a [different way](https://learn.microsoft.com/en-us/cpp/build/cmake-projects-in-visual-studio?view=msvc-170).

Instead of opening a solution file, you open the rathena folder and it will automatically use cmake to generate the project.

vstumpf commented:
Regarding file-globbing the sources list:

Even the cmake docs say they do not recommend this, as cmake won't know when to regenerate the solution because the CMakeLists.txt change signals cmake to regenerate. If we glob the sources, the file doesn't change when we add a new file and it won't regenerate: https://cmake.org/cmake/help/latest/command/file.html#glob

Also, I think it's a bad-faith argument to say that requiring a developer to add the file to the list of sources is the same as what we currently do. 

Currently, to add a new file in common, you have to modify:
- src/common/common.vcxproj
- src/common/common.vcxproj.filters
- src/common/common-minicore.vcxproj
- src/common/common-minicore.vcxproj.filters
- src/common/Makefile.in
- src/common/CMakeLists.txt

with this pr, it is just `src/common/CMakeLists.txt`

aleos89 commented:
I did finally get to play around with this PR today. It seems like a lot of functionality for CMake isn't intuitively native for MSVS. A few of the options from the solution were grayed out in this change (run, debug, etc). With this change though it will suck losing the ability to open the rAthena project directly from the root and would take some getting use to.

I can see the benefit of dropping support from us having to maintain several different compiler types. Overall, it's pretty cool. Will just take some more toying around in MSVS to see what the new shortcuts are.

vstumpf commented:
I switched to a new computer, it turns out I had to delete the `.vs` folder and reopen Visual Studio.

Then, I could switch to the CMake Targets View, select the target I want to run/debug, and go from there:

https://github.com/rathena/rathena/assets/6844975/35e9e0c9-9331-487d-90c6-ef6713feb813

vstumpf commented:
I added some presets in `CMakePresets.json` to handle the basic configuration.

People can either modify that file or create a `CMakeUserPresets.json` to create their own presets
For example, here's my preset:
```JSON
{
    "version": 6,
    "configurePresets": [
      {
        "name": "myro-debug",
        "inherits": "linux",
        "cacheVariables": {
          "CMAKE_BUILD_TYPE": "Debug",
          "ENABLE_PRERENEWAL": "ON",
          "PACKETVER": "20200401",
          "ENABLE_MEMMGR": "OFF",
          "ENABLE_EPOLL": "ON",
          "ENABLE_TESTS": "ON"
        }
      }
    ],
    "buildPresets": [
      {
        "name": "MyRO - Debug",
        "configurePreset": "myro-debug",
        "targets": ["server"]
      },
    ]
}
```

vstumpf commented:
i've added the explicit gcc/clang builds
windows by default uses msvc, we've never supported another compiler like mingw ro clang on windows, so I have feature parity there


also increased the number of build threads, as we have 4 cpus available
changed the ci cmake commands to two lines:
```
cmake -b build <configure args>
cmake --build build <build args>
```

vstumpf commented:
@Lemongrass3110 Why revert the changes to build_servers_gcc/msbuild?

Lemongrass3110 commented:
> @Lemongrass3110 Why revert the changes to build_servers_gcc/msbuild?

@vstumpf only contained changes that are out of scope of this PR.

Lemongrass3110 commented on src/common/CMakeLists.txt:
```diff
@@ -1,191 +1,230 @@
+	"timer.cpp"
+	"utils.cpp"
+	"utilities.cpp"
+)
```
```suggestion
)

if(ENABLE_EXTRA_BUILDBOT_CODE)
	file(GLOB COMMON_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/*.cpp)

	get_target_property(compile_defs common SOURCES)
	list(LENGTH compile_defs cdeflen)
	list(TRANSFORM compile_defs PREPEND ${CMAKE_CURRENT_SOURCE_DIR}/)

	foreach (my_entry IN LISTS COMMON_SOURCES)
		if (NOT my_entry IN_LIST compile_defs)
			message(FATAL_ERROR "${my_entry} is missing in target_sources for common")
		endif()
	endforeach()
endif()
```
Can we add such checks for the (most important) targets? For sources + headers.
Would require a small redesign of your common-win library though.

vstumpf replied:
Why would this be needed? If someone forgets to add the file in cmakelists.txt, then it'll fail to build.

Lemongrass3110 replied:
True, chance is really low that a header or source file will not be included in an already existing file.
May have been my way of thinking of thinking of cmake as the unmaintained compilation tool.
Since it will become our main compilation tool over time, the chances of such files will get lower and lower anyway.

Lemongrass3110 reviewed (commented):
Some more comments and findings

vstumpf commented on CMakePresets.json:
```diff
@@ -0,0 +1,93 @@
+      }
+    ]
+  }
+  
```
ran formatting on the file

vstumpf commented on .github/workflows/npc_db_validation.yml:
```diff
@@ -21,6 +21,8 @@ on:
       # This workflow should run when a file in either the db/ or npc/ directory has been modified.
       - 'db/**'
       - 'npc/**'
+      # This workflow should run whenever a CMake related file has been modified
+      - '**/CMakeLists.txt'
```
You're right, I just added this everywhere. Removing

Lemongrass3110 replied:
Thank you!

vstumpf commented on src/common/CMakeLists.txt:
```diff
@@ -1,191 +1,230 @@
+
+
+# Propagated defines
+if(ENABLE_MEMMGR)
```
What? This is on by default, this option is to just disable the memory manager.
If you want I can change it to just
```
if (NOT ENABLE_MEMGR)
	target_compile_definitions(common PUBLIC "NO_MEMMGR")
	target_compile_definitions(minicore PUBLIC "NO_MEMMGR")
endif()
```

Lemongrass3110 replied:
With the C++ migration going on and with every transition from our selfmade containers to STL containers we are working towards the full removal of the old *Athena memory manager.
Therefore I would like not to provide an option to enable/disable this feature in the new build tool.

vstumpf replied:
that makes no sense
if anything this option is better because someone can choose to disable the memory manager. right now, we're in this half-migrated state, so the option is still helpful. When we finish migrating, we can just remove the option.
Again, this is an already-existing option in `configure.ac`

vstumpf commented on src/common/CMakeLists.txt:
```diff
@@ -1,191 +1,230 @@
+	target_compile_definitions(minicore PUBLIC "-DNO_MEMMGR")
+endif()
+
+target_compile_definitions(common PUBLIC "-DPACKETVER=${PACKETVER}")
```
PACKETVER will always be set in the root `CMakeLists.txt`, by default it's `20211103`

Lemongrass3110 replied:
It should not always be set (see also comment in CMakePresets.json)
Use the default from within the code, dont duplicate defaults.

vstumpf replied:
It really should be set, the chances of someone using the "default" packetver is low. If anything, we should warn if a user doesn't pass one and we end up using the default packetver

vstumpf commented on src/common/CMakeLists.txt:
```diff
@@ -1,191 +1,230 @@
 endif()
 
+# defines just for common
+target_compile_definitions(common PRIVATE ${MONOTONIC_CLOCK_DEFINITIONS})
```
Empty defines will just be ignored by cmake
https://cmake.org/cmake/help/latest/command/target_compile_definitions.html

Lemongrass3110 replied:
Why do we have ifs around half of them then?

vstumpf replied:
Because those are just booleans, if they're set then we need to set a define. `MONOTONIC_CLOCK_DEFINITIONS` is a list of defines that will be set.

vstumpf commented on src/common/CMakeLists.txt:
```diff
@@ -1,191 +1,230 @@
+)
+
+
+if(ENABLE_PROFILER STREQUAL "gprof")
```
It's available in `configure.ac`, that's where I got it from

Lemongrass3110 replied:
Interesting. Thank you for explaining.

Lemongrass3110 commented on CMakePresets.json:
```diff
@@ -0,0 +1,93 @@
+        "hidden": true,
+        "cacheVariables": {
+          "CMAKE_EXPORT_COMPILE_COMMANDS": true,
+          "PACKETVER": "20211103"
```
We can change our documentation to add this line to CMakePresets.json or to hand the argument over to cmake, but there is no need to have it in there by default.

Since json does not allow comments and the devs of cmake made the brilliant decision to use it, we cannot even add it commented.
If you really want to have it, then please add it as a non default (maybe even hidden) with a specific name.
